lib/modules/dosomething/dosomething_api/resources/campaign_resource.inc

```diff
@@ -333,73 +333,73 @@ function _campaign_resource_signup($nid, $values) {
  * @param array $values
  *   The reportback data to post. Expected keys:
- *   - uid: The user uid (int).  Optional, uses global $user if not set.
- *   - file: Base64 encoded file string to save.
- *   - filename: The filename of the file provided as file.
- *   - file_url: The URL of the reportback file to save (used if no file/filename exist).
+ *   - uid: The user uid (int).
+ *   - file: Base64 encoded file string to save (required if using 'filename').
+ *   - filename: The filename of the file provided as file (required if using 'file').
+ *   - file_url: The URL of the reportback file to save (used if no 'file'/'filename' exist).
  *   - caption: The caption for the Reportback File if provided.
  *   - quantity (int).
  *   - why_participated (string).
  *   - num_participants (int).
  *   - remote_addr (string).
  */
 function _campaign_resource_reportback($nid, $values) {
-  // @todo: Return error if signup doesn't exist.
-
-  $values['nid'] = $nid;
+  global $user;
+  $api_user = $user;
 
-  if (isset($values['northstar_id'])) {
-    $values['uid'] = dosomething_user_get_user_by_northstar_id($values['northstar_id'])->uid;
+  if (!user_access('view any reportback', $api_user)) {
+    return dosomething_helpers_basic_http_response(401, 'Unauthorized.');
   }
-  else if (isset($values['uid'])) {
-    $user = user_load($values['uid']);
+
+  if (empty($values['uid'])) {
+    return dosomething_helpers_basic_http_response(422, 'Missing uid query parameter specifying a user id.');
   }
-  else {
-    global $user;
-    $values['uid'] = $user->uid;
+
+  $user = user_load(dosomething_user_convert_to_legacy_id($values['uid']));
+
+  if (!$user) {
+    return dosomething_helpers_basic_http_response(404, 'The specified user was not found.');
   }
 
-  $uid = $values['uid'];
+  // Set uid to member user submitting reportback.
+  $values['uid'] = $uid = $user->uid;
 
-  $file = NULL;
+  // @todo: Return error if signup doesn't exist.
 
   // @todo: Move this logic into dosomething_reportback_save().
-  $rbid = dosomething_reportback_exists($nid, NULL, $uid);
+  $rbid = dosomething_reportback_exists($nid, NULL, $uid) ?: 0;
 
-  // If this is creating a new reportback, there must be a file
+  // If file is submitted as an upload, the 'file' and 'filename' values must be provided.
   if (!empty($values['file']) && !empty($values['filename'])) {
     $values['filepath'] = dosomething_reportback_get_file_dest($values['filename'], $nid, $uid);
+
     // Use Services module's File Create resource to save the file.
     module_load_include('inc', 'services', 'resources/file_resource');
+
     $result = _file_resource_create($values);
+
     $file = file_load($result['fid']);
   }
+  elseif (!empty($values['file_url'])) {
+    $file = dosomething_reportback_save_file_from_url($nid, $uid, $values['file_url']);
 
-  if (!$file) {
-    // If this is to create a new reportback, there must be an image
-    if (!array_key_exists('file_url', $values) && !$rbid) {
```

Comment:
Oh ok. The logic felt kinda weird since it's separated from the above alternative file related params as well... I'm guessing that should also be a validation error in func that actually saves RB if it's the first RB... but ya, Drupal ;)